### PR TITLE
Add channels-last support for bilinear interpolation on CUDA

### PR DESCRIPTION
**Pull request by xwang233**
Add channels-last support for bilinear interpolation on CUDA

Related https://github.com/pytorch/pytorch/pull/48675#issuecomment-737020271 https://github.com/pytorch/pytorch/pull/54744

### CODE REVIEW

**Comment by xwang233** [comment]
reserved

**Comment by xwang233** [comment]
@ngimel any comments? 😄

File: aten/src/ATen/native/cuda/UpSampleBilinear2d.cu [hunk 1]
```diff
@@ -169,31 +292,80 @@ static void upsample_bilinear2d_out_cuda_template(
...
+        output.copy_(input);
+        return;
+      }
+
+      TORCH_CHECK(input.numel() < std::numeric_limits<int>::max(),
+        "upsample_bilinear2d only supports input tensors with less than INT_MAX elements");
+      TORCH_CHECK(output.numel() < std::numeric_limits<int>::max(),
+        "upsample_bilinear2d only supports output tensors with less than INT_MAX elements");
+
+      const int batchsize = input.size(0);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:311]: this is already computed above, see lines 287-293

File: aten/src/ATen/native/cuda/UpSampleBilinear2d.cu [hunk 2]
```diff
@@ -218,14 +390,15 @@ static void upsample_bilinear2d_backward_out_cuda_template(
...
 
-  Tensor grad_output = grad_output_.contiguous();
-
   if (grad_input.numel() == 0) {
     return;
   }
 
-  // A contiguous tensor is required for the kernel launch config
-  grad_input.contiguous();
+  const auto memory_format = grad_output_.suggest_memory_format();
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:397]: probably better to infer format from grad_input?  grad_output could have lost channels last

File: test/test_nn.py [hunk 3]
```diff
@@ -12763,6 +12763,44 @@ def helper(n, c, h, w, ks):
...
+            out.backward(g)
+            self.assertTrue(data.grad.is_contiguous())
+            out_cl.backward(g_cl)
+            self.assertTrue(data_cl.grad.is_contiguous(memory_format=memory_format_cl))
+
+            self.assertEqual(data.grad, data_cl.grad.contiguous())
+
+
+        for align_corners in [True, False]:
+            helper('bilinear', (2, 4, 16, 16), 8, align_corners)
```
Comment (ngimel) [inline test/test_nn.py:12800]: can you please add tests for non-square images?

File: test/test_nn.py [hunk 4]
```diff
@@ -12763,6 +12763,44 @@ def helper(n, c, h, w, ks):
...
         helper(None, 3, 50, 50, ks=5)
 
+    def test_interpolate_channels_last(self, device):
+        def helper(mode, input_size, out_size, align_corners):
+            memory_format_cl = torch.contiguous_format
+            ndim = len(input_size)
+            if ndim == 4:
+                memory_format_cl = torch.channels_last
+            elif ndim == 5:
+                memory_format_cl = torch.channels_last_3d
```
Comment (ngimel) [inline test/test_nn.py:12773]: please leave a comment here that 3d is not tested yet

**Comment by xwang233** [comment]
Thanks for the review @ngimel . Can you leave comments in https://github.com/pytorch/pytorch/pull/56322? It has both bilinear and nearest 2d kernels, which fully includes all contents of this PR, unless you prefer to review and merge two PRs separately 😅 .

**Comment by xwang233** [comment]
This PR is closed in favor of https://github.com/pytorch/pytorch/pull/56322

**Pull request closed without merging**